internal/service/ipam/api_next_available_ip_data_source.go

```diff
@@ -98,54 +107,285 @@ func (d *NextAvailableIPDataSource) Schema(_ context.Context, _ datasource.Schem
 				Computed:            true,
 				MarkdownDescription: "List of next available IP addresses in the specified resource",
 			},
+			"tag_filters": schema.MapAttribute{
+				ElementType:         types.StringType,
+				Optional:            true,
+				MarkdownDescription: "Map of tag key/value pairs to filter resources",
+				Validators: []validator.Map{
+					mapvalidator.AlsoRequires(path.MatchRoot("resource_type")),
+				},
+			},
+			"resource_type": schema.StringAttribute{
+				Optional:            true,
+				MarkdownDescription: "Resource type to search when using tag filters (address_block, subnet, or range)",
+				Validators: []validator.String{
+					stringvalidator.OneOf("address_block", "subnet", "range"),
+					stringvalidator.AlsoRequires(path.MatchRoot("tag_filters")),
+				},
+			},
 		},
 	}
 }
 
 func (d *NextAvailableIPDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
-	var (
-		data   IpamsvcNextAvailableIPModel
-		apiRes *ipam.NextAvailableIPResponse
-		err    error
-	)
+	var data IpamsvcNextAvailableIPModel
 
 	// Read Terraform prior state data into the model
 	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
 
 	if resp.Diagnostics.HasError() {
 		return
 	}
-	addressStr := data.Id.ValueString()
 
-	switch addressStr[:strings.LastIndex(addressStr, "/")] {
+	// Set default count if not provided
+	count := int64(1)
+	if !data.Count.IsNull() {
+		count = data.Count.ValueInt64()
+	}
+
+	// Validate count is within allowed range (1-20)
+	if count < 1 || count > 20 {
+		resp.Diagnostics.AddError("Invalid count", "Count must be between 1 and 20")
+		return
+	}
+
+	// Check if we're using tag_filters or direct resource ID
+	if !data.TagFilters.IsNull() {
+		// Using tag filters - ensure resource_type is specified
+		if data.ResourceType.IsNull() {
+			resp.Diagnostics.AddError("Missing resource_type", "resource_type is required when using tag_filters")
+			return
+		}
+
+		// Get resources by tags and find next available IPs
+		addresses, err := d.findNextAvailableIPsByTags(ctx, data)
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to find resources by tags: %s", err))
+			return
+		}
+
+		if len(addresses) < int(count) {
+			resp.Diagnostics.AddError("Insufficient IPs", fmt.Sprintf("Not enough available IPs found in %s with the given tags", data.ResourceType.ValueString()))
+			return
+		}
+
+		// Limit to requested count
+		if len(addresses) > int(count) {
+			addresses = addresses[:count]
+		}
+
+		data.FlattenResults(ctx, addresses, &resp.Diagnostics)
+	} else if !data.Id.IsNull() {
+		// Using direct resource ID
+		addressStr := data.Id.ValueString()
+		apiRes, err := d.getNextAvailableIPsByID(ctx, addressStr, count, data.Contiguous.ValueBool())
```

Comment:
```suggestion
		apiRes, err := d.getNextAvailableIPsByID(ctx, data.Id.ValueString(), count, data.Contiguous.ValueBool())
```

There is only one usage for `data.Id.ValueString()` , variable definition isnt required !